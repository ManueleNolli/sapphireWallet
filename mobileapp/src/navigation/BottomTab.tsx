import React, { useEffect, useRef } from 'react'
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs'
import {
  BottomNavigation,
  BottomNavigationTab,
  Icon,
  IconElement,
} from '@ui-kitten/components'
import { Image } from 'expo-image'
import { logo } from '../assets/AssetsRegistry'

// Pages
import Home from '../pages/Home/Home'
import Details from '../pages/Details/Details'
import Settings from '../pages/Settings/Settings'
import { Button, ImageProps, View } from 'react-native'

const { Navigator, Screen } = createBottomTabNavigator()

const BottomTabBar = ({ navigation, state }: any) => {
  const homeIconRef = useRef<Icon<Partial<ImageProps>>>()
  const NFTsIconRef = useRef<Icon<Partial<ImageProps>>>()
  const settingsIconRef = useRef<Icon<Partial<ImageProps>>>()

  const HomeIcon = (props: any): IconElement => (
    <Icon
      {...props}
      ref={homeIconRef}
      animation="pulse"
      name="person"
      animationConfig={{
        duration: 200,
      }}
    />
  )
  const NFTsIcon = (props: any): IconElement => (
    <Icon
      {...props}
      ref={NFTsIconRef}
      animation="pulse"
      name="image"
      animationConfig={{
        duration: 200,
      }}
    />
  )
  const SettingsIcon = (props: any): IconElement => (
    <Icon
      {...props}
      ref={settingsIconRef}
      animation="pulse"
      name="settings"
      animationConfig={{
        duration: 200,
      }}
    />
  )

  useEffect(() => {
    if (state.index === 0 && homeIconRef.current) {
      homeIconRef.current.startAnimation()
    } else if (state.index === 1 && NFTsIconRef.current) {
      NFTsIconRef.current.startAnimation()
    } else if (state.index === 2 && settingsIconRef.current) {
      settingsIconRef.current.startAnimation()
    }
  }, [state.index])

  return (
    <BottomNavigation
      selectedIndex={state.index}
      onSelect={(index) => navigation.navigate(state.routeNames[index])}
    >
      <BottomNavigationTab title="Balance" icon={HomeIcon} />
      <BottomNavigationTab title="NFTs" icon={NFTsIcon} />
      <BottomNavigationTab title="Settings" icon={SettingsIcon} />
    </BottomNavigation>
  )
}

function LogoIcon() {
  return (
    <Image
      contentFit={'contain'}
      style={{
        width: 32,
        height: 32,
        marginLeft: 8,
      }}
      source={logo}
    />
  )
}

function NetworkIcon() {
  return (
    <Image
      contentFit={'contain'}
      style={{
        width: 32,
        height: 32,
        marginRight: 8,
      }}
      source={logo}
    />
  )
}

const BottomTabNavigator = () => (
  <Navigator
    screenOptions={{ headerTransparent: true, headerTitleAlign: 'center' }}
    tabBar={(props) => <BottomTabBar {...props} />}
  >
    <Screen
      name="Home"
      component={Home}
      options={({ navigation, route }) => ({
        headerLeft: () => <LogoIcon />,
        headerRight: () => <NetworkIcon />,
      })}
    />
    <Screen name="Details" component={Details} />
    <Screen name="Settings" component={Settings} />
  </Navigator>
)

export default BottomTabNavigator
