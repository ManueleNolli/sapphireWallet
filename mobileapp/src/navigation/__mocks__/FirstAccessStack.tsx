const FirstAccessStackMock = () => (
  <div data-testid={'FirstAccessStack'}>FirstAccessStackMock</div>
)

export default FirstAccessStackMock
