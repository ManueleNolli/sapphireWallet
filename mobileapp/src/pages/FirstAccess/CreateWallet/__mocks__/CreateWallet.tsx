const CreateWalletMock = () => (
  <div data-testid={'CreateWallet'}>CreateWallet</div>
)

export default CreateWalletMock
