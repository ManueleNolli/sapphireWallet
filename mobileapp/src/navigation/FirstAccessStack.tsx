import {
  CardStyleInterpolators,
  createStackNavigator,
  StackScreenProps,
} from '@react-navigation/stack'
import CreateWallet from '../pages/FirstAccess/CreateWallet/CreateWallet'
import MnemonicViewer from '../pages/FirstAccess/MnemonicViewer/MnemonicViewer'

///////////
// TYPES //
///////////
/**
 * FirstAccessStackParamList is a type that represents the possible routes
 * If undefined, it means that the route doesn't have any params
 */
type FirstAccessStackParamList = {
  CreateWallet: undefined
  MnemonicViewer: {
    mnemonic: string[]
  }
}

export type MnemonicViewerProps = StackScreenProps<
  FirstAccessStackParamList,
  'MnemonicViewer'
>

export type CreateWalletProps = StackScreenProps<
  FirstAccessStackParamList,
  'CreateWallet'
>

////////////////
// NAVIGATION //
////////////////

const { Navigator, Screen } = createStackNavigator<FirstAccessStackParamList>()

const FirstAccessStackNavigator = () => (
  <Navigator
    initialRouteName="CreateWallet"
    screenOptions={{
      headerShown: false,
      gestureEnabled: true,
      cardStyleInterpolator: CardStyleInterpolators.forHorizontalIOS,
    }}
  >
    <Screen name="CreateWallet" component={CreateWallet} />
    <Screen name="MnemonicViewer" component={MnemonicViewer} />
  </Navigator>
)
export default FirstAccessStackNavigator
