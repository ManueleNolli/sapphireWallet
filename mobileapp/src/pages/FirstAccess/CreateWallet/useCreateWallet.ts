import { CreateWalletProps } from '../../../navigation/FirstAccessStack'
import { createWallet, getMnemonic } from '../../../services/wallet'
import { useContext } from 'react'
import { WalletContext } from '../../../context/WalletContext'

export default function useCreateWallet(
  navigation: CreateWalletProps['navigation']
) {
  const { setPrivateKey, setEOAAddress } = useContext(WalletContext)

  const createAndNavigate = async () => {
    const wallet = createWallet()
    try {
      await Promise.all([
        setEOAAddress(wallet.address),
        setPrivateKey(wallet.privateKey),
      ])

      navigation.push('MnemonicViewer', { mnemonic: getMnemonic(wallet) })
    } catch (errors) {
      console.error(errors)
    }
  }

  return {
    createAndNavigate,
  }
}
