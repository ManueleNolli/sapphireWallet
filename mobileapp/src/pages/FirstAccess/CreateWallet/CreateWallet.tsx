import React from 'react'

import { Button, Layout } from '@ui-kitten/components'
import { View } from 'react-native'
import { appStyles, vh } from '../../../Styles'
import { Image } from 'expo-image'
import { logoWithFullText } from '../../../assets/AssetsRegistry'
import useCreateWallet from './useCreateWallet'
import SafeAreaView from '../../../utils/SafeAreaView'
import { CreateWalletProps } from '../../../navigation/FirstAccessStack'

export default function CreateWallet({ navigation }: CreateWalletProps) {
  const { createAndNavigate } = useCreateWallet(navigation)

  return (
    <Layout style={{ flex: 1 }}>
      <SafeAreaView style={{ marginVertical: 20 * vh }}>
        <View style={{ flex: 1 }}>
          <Image style={appStyles.imageContain} source={logoWithFullText} />
        </View>
        <View style={[{ flex: 1 }, appStyles.centerNoFlex]}>
          <Button
            appearance="outline"
            size={'giant'}
            onPress={createAndNavigate}
          >
            Create wallet
          </Button>
        </View>
      </SafeAreaView>
    </Layout>
  )
}
