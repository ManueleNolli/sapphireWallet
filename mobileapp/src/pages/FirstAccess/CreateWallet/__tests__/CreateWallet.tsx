import renderWithTheme from '../../../../TestHelper'
import CreateWallet from '../CreateWallet'
import { CreateWalletProps } from '../../../../navigation/FirstAccessStack'
import useCreateWallet from '../useCreateWallet'

// MOCKS
jest.mock('../useCreateWallet', () => jest.fn())

const props = {
  navigation: {
    navigate: jest.fn(),
  },
  route: jest.fn(),
} as unknown as CreateWalletProps

describe('CreateWallet', () => {
  beforeEach(() => {
    ;(useCreateWallet as jest.Mock).mockReturnValue({
      createAndNavigate: jest.fn(),
    })
  })

  it('Render correctly', () => {
    const tree = renderWithTheme(<CreateWallet {...props} />)

    expect(tree).toMatchSnapshot()
  })
})
