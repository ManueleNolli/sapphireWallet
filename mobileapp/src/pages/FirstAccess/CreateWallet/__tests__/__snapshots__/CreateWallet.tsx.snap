// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateWallet Render correctly 1`] = `
<View
  appearance="default"
  level="1"
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
      },
      {
        "flex": 1,
      },
    ]
  }
>
  <View
    appearance="default"
    level="1"
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
        },
        {
          "flex": 1,
        },
      ]
    }
  >
    <RNCSafeAreaView
      edges={
        {
          "bottom": "additive",
          "left": "additive",
          "right": "additive",
          "top": "additive",
        }
      }
      style={
        [
          {
            "flex": 1,
            "marginHorizontal": 37.5,
            "marginVertical": 66.7,
          },
          {
            "marginVertical": 266.8,
          },
        ]
      }
    >
      <View
        style={
          {
            "flex": 1,
          }
        }
      >
        <ViewManagerAdapter_ExpoImage
          contentFit="contain"
          contentPosition={
            {
              "left": "50%",
              "top": "50%",
            }
          }
          flex={1}
          onError={[Function]}
          onLoad={[Function]}
          onLoadStart={[Function]}
          onProgress={[Function]}
          placeholder={[]}
          source={[]}
          style={
            {
              "contentFit": "contain",
              "flex": 1,
            }
          }
          transition={null}
        />
      </View>
      <View
        style={
          [
            {
              "flex": 1,
            },
            {
              "alignItems": "center",
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(51, 102, 255, 0.08)",
              "borderColor": "#3366FF",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "justifyContent": "center",
              "minHeight": 56,
              "minWidth": 56,
              "opacity": 1,
              "paddingHorizontal": 12,
              "paddingVertical": 16,
            }
          }
        >
          <Text
            appearance="default"
            category="p1"
            status="basic"
            style={
              [
                {
                  "color": "#222B45",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "400",
                },
                {
                  "color": "#3366FF",
                  "fontFamily": "System",
                  "fontSize": 18,
                  "fontWeight": "bold",
                  "marginHorizontal": 12,
                },
              ]
            }
          >
            Create wallet
          </Text>
        </View>
      </View>
    </RNCSafeAreaView>
  </View>
</View>
`;
